dags/trending-projects-snapshot-dag.py

```diff
@@ -0,0 +1,183 @@
+from airflow.models.param import Param
+from airflow.providers.snowflake.hooks.snowflake import SnowflakeHook
+from orca.services.synapse import SynapseHook
+
+
+SYNAPSE_RESULTS_TABLE = "syn61597055"
+
+dag_params = {
+    "snowflake_conn_id": Param("SNOWFLAKE_SYSADMIN_PORTAL_RAW_CONN", type="string"),
+    "synapse_conn_id": Param("SYNAPSE_ORCA_SERVICE_ACCOUNT_CONN", type="string"),
+    "current_date": Param(date.today().strftime("%Y-%m-%d"), type="string"),
+    "time_window": Param("30", type="string"),
+    }
+
+dag_config = {
+    # run on the 2nd of every month at midnight
+    "schedule_interval": "0 0 2 * *",
+    "start_date": datetime(2024, 1, 1),
+    "catchup": False,
+    "default_args": {
+        "retries": 1,
+    },
+    "tags": ["snowflake"],
+    "params": dag_params,
+}
+
+
+@dataclass
+class SnapshotMetrics:
+    """
+    Dataclass to hold the download metrics from Synapse.
+
+    Attributes:
+        project_id: The ID of the Project
+        n_unique_users: The number of unique users who downloaded from the Project
+        last_download_date: The date of the last download from the Project
+        total_data_size_in_gib: The current total size of the Project in GiB
+    """
+
+    project_id: float
+    n_unique_users: int
+    last_download_date: int
+    total_data_size_in_gib: float
+
+
+@dag(**dag_config)
+def trending_projects_snapshot() -> None:
+    """
+    This DAG executes a query on Snowflake to retrieve information about trending public projects and 
+    reports the results to a Synapse table.
+
+    The main steps performed in this DAG are:
+    1. Query Snowflake to identify public projects, their file sizes, and recent download activities.
+    2. Aggregate the results to determine num. of unique users, last download date, and total data size.
+    3. Format the results into a list of SnapshotsMetric objects.
+
+    The Snowflake connection and Synapse connection IDs, as well as the time window for recent downloads,
+    are provided via DAG parameters.
+    """
+
+    @task
+    def get_trending_project_snapshot(**context) -> List[SnapshotMetrics]:
+        """Execute the query on Snowflake and return the results."""
+        snow_hook = SnowflakeHook(context["params"]["snowflake_conn_id"])
+        ctx = snow_hook.get_conn()
+        cs = ctx.cursor()
+        query = f"""
+                WITH PUBLIC_PROJECTS AS (
+                    SELECT PROJECT_ID
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.NODE_LATEST
+                    WHERE 1=1
+                    AND NODE_TYPE = 'project'
+                    AND IS_PUBLIC = TRUE
+                ),
+                LATEST_FILE_HANDLES AS (
+                    SELECT DISTINCT ID, PROJECT_ID, FILE_HANDLE_ID
+                    FROM SYNAPSE_DATA_WAREHOUSE.SYNAPSE.NODE_LATEST
+                    WHERE 1=1
+                    AND NODE_TYPE = 'file'
+                    AND PROJECT_ID IN (SELECT PROJECT_ID FROM PUBLIC_PROJECTS)
```

Comment:
NIt: I was going to say you could combine the PUBLIC_PROJECTS / LATEST_FILE_HANDLES / FILE_SIZES into one CTE, but I understand why you have PUBLIC_PROJECTS as a separate CTE.

Reply:
While I was working on the optimization suggestion from [this thread](https://github.com/Sage-Bionetworks-Workflows/orca-recipes/pull/65#discussion_r1701338028), I considered having the `RECENT_DOWNLOADS` CTE join `FILEDOWNLOAD` and `NODE_LATEST` and getting rid of `PUBLIC_PROJECTS` altogether, but I decided against it because:

- I don't believe it would minimize compute time
- Smaller CTEs are easier to read and makes the query easier to understand (from my perspective)

But lmk what you think!